Feat: export user & extension defined properties

Necronomison opened the pull request:
I agree that my contributions are licensed under the Individual Contributor License Agreement V4.0 ("CLA") as stated in https://github.com/igarastudio/cla/blob/main/cla.md

I have signed the CLA following the steps given in https://github.com/igarastudio/cla#signing

-----------------------------------------------------

### Related Issue(s):
Resolves https://github.com/aseprite/aseprite/issues/5186

### Context:
Since [Aseprite 1.3-rc1](https://aseprite.org/api/properties#properties) you can have user-defined and extension-defined properties for each of these objects. 

However, these properties are currently not exported into the JSON file.

This would be a very handy feature so that plugins can fully take advantage of these custom properties when parsing aseprite json files in other tools (Unity, etc...).

### Solution:

- **Full serialization of custom properties for Tags, Layers, Cels, and Slices:**  
  The exported JSON now includes all user-defined and extension-defined properties for these objects, under a `"properties"` key. This ensures that any metadata added by users or extensions is preserved and accessible in downstream tools.
- **Generic, type-safe serialization helper:**  
  Introduced a reusable, type-constrained helper function (`serialize_userdata_properties`) that serializes the properties of any object derived from `WithUserData`. This ensures consistent and safe handling of properties across all supported object types.
- **Accurate variant type handling:**  
  The serialization logic for property values robustly supports all integer types and other supported types in `UserData::Variant`, guaranteeing correct JSON output for all property values, regardless of how they are stored or loaded.



### Testing:

#### Compilation:
- [x] Control compilation at base `master` (SHA `66123e9d5739fdf870336fd4eda91e0054421ea0`)
- [x] Compilation at tip of this feature branch (SHA `42e386b701064b9e8a4de9409ec7cbf39f40a7e4`)

#### Functional Test Cases:
- [x] Frame Tags
- [x] Layer
- [x] Cel
- [x] Slice

In each case, a similar lua snippet was run to add a user and group/extension property to each of the above objects.
```
function injectPropertiesFor(targetType)
  local target = app[targetType]
  if not target then
    app.alert("No active " .. targetType .. ".")
    return
  end
  target.properties = {}
  target.properties.user_defined_data = targetType .. " user defined data"
  local props = target.properties("my-plugin")
  if not props then
    target.properties("my-plugin", {})
    props = target.properties("my-plugin")
  end
  props.extension_defined_data = targetType .. " extension defined data"
  app.alert("Injected " .. targetType .. " properties.")
end
```

JSON Result:
[meta.json](https://github.com/user-attachments/files/20544825/meta.json)

_(Note that the meta data has been manually vertically restructured just for the purpose of better visibility in this MR. I haven't changed any logic for how it formats in the actual exported JSON (flat/wide))_

WolvenBard commented:
@dacap I've addressed the macOS 10.9 C++ runtime incompatibility for my `serialize_variant` method in the latest commit. 

Would we be able to re-run the workflows? Or is there a permission setting I can apply for to rerun them myself?

Many thanks :)

WolvenBard commented:
@dacap, it looks like another 2 workflows need to be kicked off? The other ones are green now though.

dacap commented on src/app/doc_exporter.cpp:
```diff
@@ -69,6 +69,106 @@ std::string escape_for_json(const std::string& path)
+void serialize_properties(const doc::UserData::Properties& props, std::ostream& os)
+{
+  os << "{";
+  bool first = true;
```
I don't know why clang-tidy-diff fails doing this commenting so much about local variables that can be const when they cannot. Just ignore this.

Necronomison commented:
> Hi @Necronomison, PRs got merged when we have time to review them. I'll try to review this probably this Monday. We're planning a release next week. A couple of comments:
> 
> 1. From what I see, it looks like `serialize_userdata_properties()` should be part of/called from the `std::ostream& operator<<(std::ostream& os, const doc::UserData& data)` function.
> 2. Would you mind to add some tests about this? Probably in a new `tests/scripts/userdata_json.lua` (something similar to `tests/scripts/userdata_codec.lua`)?


Awesome, thank you for the information @dacap.

I've updated my merge request with the following changes:
- Moved the call to my `serialize_properties` from the UserData stream function.
- Added a new `userdata_json.lua` and tested that the output and asserts/cases are green
- Fixed a bit of json formatting so that there is a space after after comma separators in the properties lists

Looking forward to any other review requests you may have!

Necronomison commented on src/app/doc_exporter.cpp:
```diff
@@ -69,6 +69,104 @@ std::string escape_for_json(const std::string& path)
+{
+  os << "{";
+  bool first = true;
+  for (const auto& [key, value] : props) {
```
Closing this as its the same (false) suggestion as this:
https://github.com/aseprite/aseprite/pull/5187#discussion_r2123577525

Necronomison commented on src/app/doc_exporter.cpp:
```diff
@@ -69,6 +69,104 @@ std::string escape_for_json(const std::string& path)
+{
+  const auto& propsMaps = data.propertiesMaps();
+  bool hasAnyProps = false;
+  for (const auto& [group, props] : propsMaps) {
```
Closing this as its the same (false) suggestion as this:
https://github.com/aseprite/aseprite/pull/5187#discussion_r2123577525

Necronomison commented:
@dacap fixed

dacap commented:
Sorry @Necronomison, I wasn't able to check the latest email about the CLA yet (finishing the v1.3.14 release here).

dacap commented:
Thanks for this @Necronomison, I'll merge this squashed right now for the next release v1.3.14.2 which might be tomorrow. I'll just modify the test output dir:

https://github.com/aseprite/aseprite/pull/5187/files#diff-6d829e153c5f8ba2992d80d1a5030aa9b5aab774795e8b23804be0b50a235347R6-R12

As we shouldn't use the user config directory to output files from automated tests.
